packages/cosmic-swingset/src/launch-chain.js

```diff
@@ -790,146 +884,28 @@ export async function launch({
       case ActionType.AG_COSMOS_INIT: {
-        const { isBootstrap, blockTime, params } = action;
-        let upgradePlan = action.upgradePlan;
-        const { coreProposals: upgradeCoreProposals, info: upgradePlanInfo } =
-          parseCoreProposalsFromUpgradePlan(
-            upgradePlan,
-            `${ActionType.AG_COSMOS_INIT} `,
-          );
-
-        // This only runs for the very first block on the chain.
-        if (isBootstrap) {
-          verboseBlocks && blockManagerConsole.info('block bootstrap');
-          (savedHeight === 0 && savedBeginHeight === 0) ||
-            Fail`Cannot run a bootstrap block at height ${savedHeight}`;
-          const bootstrapBlockParams = parseParams(params);
-          const blockHeight = 0;
-          controller.writeSlogObject({
-            type: 'cosmic-swingset-bootstrap-block-start',
-            blockTime,
-          });
-          // Start a block transaction, but without changing state
-          // for the upcoming begin block check
-          saveBeginHeight(savedBeginHeight);
-          await processAction(action.type, async () =>
-            bootstrapBlock(blockHeight, blockTime, bootstrapBlockParams),
-          );
+        const { blockHeight, isBootstrap, upgradePlan } = action;
 
-          // Merge the core proposals from the bootstrap block with the
-          // ones from the upgrade plan.
-          const coreProposals = mergeCoreProposals(
-            bootstrapCoreProposals,
-            upgradeCoreProposals,
-          );
-
-          // Synthesize an upgrade plan for the bootstrap block.
-          upgradePlan = harden({
-            ...upgradePlan,
-            info: JSON.stringify({
-              ...upgradePlanInfo,
-              coreProposals,
-            }),
-            height: blockHeight,
-          });
-        }
-        controller.writeSlogObject({
-          type: 'cosmic-swingset-bootstrap-block-finish',
-          blockTime,
-        });
-        if (upgradePlan) {
-          const blockHeight = upgradePlan.height;
+        let { coreProposals } = parseUpgradePlanInfo(
+          upgradePlan,
+          ActionType.AG_COSMOS_INIT,
+        );
 
-          // Process upgrade plan
-          const upgradedAction = {
-            type: ActionType.ENACTED_UPGRADE,
-            upgradePlan,
-            blockHeight,
-            blockTime,
-          };
-          await doBlockingSend(upgradedAction);
+        if (isBootstrap) {
+          // This only runs for the very first block on the chain.
+          coreProposals = await doCosmosInit(action, coreProposals);
         }
-        return true;
-      }
-
-      case ActionType.ENACTED_UPGRADE: {
-        // Install and execute new core proposals.
-        const { upgradePlan, blockHeight, blockTime } = action;
 
-        if (!blockNeedsExecution(blockHeight)) {
-          return undefined;
+        if (!coreProposals || !blockNeedsExecution(blockHeight)) {
+          return true;
```

Comment:
While correct, this condition is confusing me. Can we restore the original, and instead do the following below?
```js
        if (coreProposals) {
          await doCoreProposals(action, coreProposals);
        }
        return true;
```